esp-hal/src/dma/mod.rs

```diff
@@ -1840,6 +1878,519 @@ where
+        for (desc, chunk) in chunk_iter {
+            desc.set_size(chunk.len());
+            desc.buffer = chunk.as_mut_ptr();
+        }
+
+        let mut buf = Self {
+            descriptors,
+            buffer,
+        };
+        buf.set_length(buf.capacity());
+
+        Ok(buf)
+    }
+
+    /// Consume the buf, returning the descriptors and buffer.
+    pub fn split(self) -> (&'static mut [DmaDescriptor], &'static mut [u8]) {
+        (self.descriptors, self.buffer)
+    }
+
+    /// Returns the size of the underlying buffer
+    pub fn capacity(&self) -> usize {
+        self.buffer.len()
+    }
+
+    /// Return the number of bytes that would be transmitted by this buf.
+    #[allow(clippy::len_without_is_empty)]
+    pub fn len(&self) -> usize {
+        let mut result = 0;
+        for desc in self.descriptors.iter() {
+            result += desc.len();
+            if desc.next.is_null() {
+                break;
+            }
+        }
+        result
+    }
+
+    /// Reset the descriptors to only transmit `len` amount of bytes from this
+    /// buf.
+    ///
+    /// The number of bytes in data must be less than or equal to the buffer
+    /// size.
+    pub fn set_length(&mut self, len: usize) {
+        assert!(len <= self.buffer.len());
+
+        // Get the minimum number of descriptors needed for this length of data.
+        let descriptor_count = len.div_ceil(CHUNK_SIZE).max(1);
+        let required_descriptors = &mut self.descriptors[0..descriptor_count];
+
+        // Link up the relevant descriptors.
+        let mut next = core::ptr::null_mut();
+        for desc in required_descriptors.iter_mut().rev() {
+            desc.next = next;
+            next = desc;
+        }
+
+        let mut remaining_length = len;
+        for desc in required_descriptors.iter_mut() {
+            // As this is a simple dma buffer implementation we won't
+            // be making use of this feature.
+            desc.set_suc_eof(false);
+
+            // This isn't strictly needed for this simple implementation,
+            // but it is useful for debugging.
+            desc.set_owner(Owner::Dma);
+
+            let chunk_size = min(remaining_length, desc.flags.size() as usize);
+            desc.set_length(chunk_size);
+            remaining_length -= chunk_size;
+        }
+        debug_assert_eq!(remaining_length, 0);
+
+        required_descriptors.last_mut().unwrap().set_suc_eof(true);
+    }
+
+    /// Fills the TX buffer with the bytes provided in `data` and reset the
+    /// descriptors to only cover the filled section.
+    ///
+    /// The number of bytes in data must be less than or equal to the buffer
+    /// size.
+    pub fn fill(&mut self, data: &[u8]) {
+        self.set_length(data.len());
+        self.as_mut_slice()[..data.len()].copy_from_slice(data);
+    }
+
+    /// Returns the buf as a mutable slice than can be written.
+    pub fn as_mut_slice(&mut self) -> &mut [u8] {
```

Comment:
In its current form this function is a footgun. It gives access to the whole buffer, but it does not update the length.

Should we
- ... set the maximum length when this is called?
- ... only return a subset of the buffer, controlled by `set_length`?
- ... document that the user should probably call `set_length`?

Reply:
I agree that this is confusing, which is why I didn't implement `Deref`/`AsRef`/`Borrow`.

I know I definitely want users to be able to access the whole buffer regardless of length somehow, how this is done I'm not too fussed about.

Definitely not the first option at least. Since there will be no way to edit the buffer without changing its length.

This is issue worthy



> In its current form this function is a footgun.

I hope you didn't shoot your foot recently :sweat_smile:

Reply:
Some more options:

* A closure based API where the user has to return the number of bytes "used".
* We remove `as_mut_slice` and rely on `fill` only.
* We consider DmaTxBuf etc short lived, slice manipulation should happen outside before passing it into DmaTxBuf